blog/_posts/2024-10-15-scala-3.5.0-released.md

```diff
@@ -0,0 +1,119 @@
+
+### `var` in refinements
+
+Until now, only types, `val`s, and `def`s were allowed in type refinements. In 3.5.0, `var`s can also be declared.
+
+```scala
+type A = { var number: Int }
+```
+
+is now legal and equivalent to
+
+```scala
+type A = { def number: Int, def number_=($number: Int): Unit }
+```
+
+This change can simplify libraries based on metaprogramming using type refinements.
+
+### Support for binary integer literals
+
+Integer literals can now start with `0b` or `0B`. They will be interpreted as numbers in base 2.
+
+```scala
+assert(0B1 == 1)
+assert(0B10 == 2)
+assert(0B_1000_0010 == 130)
+assert(0B_1000_0010 == 0x82)
+```
+
+## Work on a better scheme for given prioritization
+
+Givens in Scala 3 have a peculiar problem with prioritization. To illustrate it, let's consider the following inheritance triangle for type classes:
+
+```scala
+trait Functor[F[_]]:
+  extension [A, B](x: F[A]) def map(f: A => B): F[B]
+trait Monad[F[_]] extends Functor[F] { ... }
+trait Traverse[F[_]] extends Functor[F] { ... }
+```
+
+and corresponding instances
+
+```scala
+given Functor[List]  = ???
+given Monad[List]    = ???
+given Traverse[List] = ???
+```
+
+Now, let's use the instances in the following context:
+
+```scala
+def fmap[F[_] : Functor, A, B](c: F[A])(f: A => B): F[B] = c.map(f)
+fmap(List(1,2,3))(_.toString) // ERROR
+```
+
+The compiler will reject the above code with a message about the `Functor` context bound being ambiguous. This is unintuitive and undesired behavior. The reason for this is that `Monad` and `Traverse` types are subtypes of `Functor`, so their instances are more specific than the `Functor` instance. They, in turn, are incomparable in terms of specificity, so the result is ambiguity.
+
+The solution to this kind of problem is to change the scheme of given prioritization. If we decide that we are always selecting the instance with **the most general subtype** that satisfies the context bound, the above example would work as intended.
+
+We have conducted experiments that showed that the proposed scheme will result in a more intuitive and predictable given resolution. The negative impact on the existing projects is very small. We have tested 1500 open-source libraries, and new rules are causing problems for less than a dozen of them. We have already submitted PRs with changes that will make them work the same way under both the current and proposed rules.
+
+Our current plan is to introduce the new scheme in Scala 3.7. Starting from Scala 3.6, code whose behavior can differ between new and old rules (ambiguity on new, passing on old, or vice versa) will emit warnings, but the old rules will still be applied. 3.5 gives you a chance to detect if those changes affect your codebase. Running the compiler with `-source 3.6` will give you warnings; with `-source 3.7` or `-source future` you will get the new scheme.
+
```

Comment:
TODO: link to the blog post once it's available (https://github.com/scala/scala-lang/pull/1675)

Reply:
another possibility would be to greatly shorten this section and let the blog post do most of the work

Reply:
I recently heard from @bracevac that the post will be closer to 3.6. This is why I needed to include the information about the planned changes right now.
However, I agree that it will be better to slightly change the post and release it now. Then we can shorten this section considerably.